keygen/src/ssl.c

```diff
@@ -0,0 +1,117 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <openssl/evp.h>
+#include <openssl/rsa.h>
+#include <openssl/pem.h>
+#include <openssl/rand.h>
+#include <openssl/ecdsa.h>
+#include <openssl/x509.h>
+
+#include "ssl.h"
+
+/**
+ * gen_EC() - Generate an Elliptic Curve key (NID_X9_62_prime256v1)
+ *
+ * Return: A pointer to a new EC_KEY instance or NULL on failure
+ */
+EC_KEY * gen_EC()
+{
+    EC_KEY * ec_key = EC_KEY_new_by_curve_name(NID_X9_62_prime256v1);
+
+    if (ec_key && !EC_KEY_generate_key(ec_key))
+    {
+        EC_KEY_free(ec_key);
+        ec_key = NULL;
+    }
+
+    return ec_key;
+}
+
+/**
+ * gen_cert() - Generate an X509 cert with a public key
+ * @pkey: The public key to certify as an EVP_PKEY*
+ * @cn: The IPv6 address (as a string) to assign to CN
+ *
+ * Return: A pointer to a new X509 certificate
+ */
+X509 *gen_cert(EVP_PKEY *pkey, char *cn)
+{
+    X509 *x509 = X509_new();
+
+    if(!x509)
+    {
+        goto out;
+    }
+
+    /* Set the serial number. */
+    ASN1_INTEGER_set(X509_get_serialNumber(x509), 1);
+
+    /* This certificate is valid from now until exactly 365 days from now */
+    X509_gmtime_adj(X509_get_notBefore(x509), 0);
+    X509_gmtime_adj(X509_get_notAfter(x509), 31536000L);
+
+    /* Set the public key for our certificate. */
+    X509_set_pubkey(x509, pkey);
+
+    /* We want to copy the subject name to the issuer name. */
+    X509_NAME * name = X509_get_subject_name(x509);
+
+    /* Set the country code and common name. */
+    X509_NAME_add_entry_by_txt(name, "C",  MBSTRING_ASC, (unsigned char *)"US",        -1, -1, 0);
+    X509_NAME_add_entry_by_txt(name, "O",  MBSTRING_ASC, (unsigned char *)"Xaptum", -1, -1, 0);
+    X509_NAME_add_entry_by_txt(name, "CN", MBSTRING_ASC, (unsigned char *)cn, -1, -1, 0);
+
+    /* Now set the issuer name. */
+    X509_set_issuer_name(x509, name);
+
+out:
+    return x509;
+}
+
+/**
+ * write_cert() - Write an X509 cert to a file
+ * @cert: The certificate to write
+ * @fname: A string containing the name of the file to write into
+ *
+ * Return: None
+ */
+void write_cert(X509 *cert, char *fname)
+{
+    FILE *f;
+    f = fopen(fname, "wb");
+
+    if (f) {
+        PEM_write_X509(f, cert);
+        fclose(f);
+    }
+}
+
+/**
+ * write_cert() - Write a puiblic and private key in PEM format
+ * @ec_key: The key to be written
+ * @fname: A string containing the name of the file to write into
+ *
+ * Return: None
+ */
+void save_EC_private(EVP_PKEY *pkey, char *fname)
+{
+    FILE *f = fopen(fname, "wb");
+    PEM_write_PrivateKey(f, pkey, NULL,NULL,0,NULL,NULL);
```

Comment:
I'm not sure how you're planning on calling `save_EC_privatekey`, but the output of the `gen_EC` (a `EC_KEY` structure) has to be converted into a `PKEY` structure before calling `PEM_write_PrivateKey`.

So I think you'll need code like
```
    EVP_PKEY *pkey = EVP_PKEY_new();
    if (!pkey) {
        <handle error>
    }   

    if (!EVP_PKEY_assign_EC_KEY(pkey, ec_key)) {
        <handle error>
    }   
```

It looks like OpenSSL does have a ecc-specific write function, but it's only v1.1 (so, now that we're supporting v1.0, it's probably easier to just convert to a PKEY no matter the version)

Reply:
I had initially thought the caller would need both the key and pkey, but that doesn't appear to be the case so I'm just going to be passing an `EVP_PKEY *` around. The docs say that when the pkey is free'd it will also free the underlying key.